nixos/modules/services/web-apps/mobilizon.nix

```diff
@@ -0,0 +1,358 @@
+      description = "Mobilizon setup secrets";
+      before = [ "mobilizon.service" ];
+      wantedBy = [ "mobilizon.service" ];
+
+      script =
+        let
+          # Taken from here:
+          # https://framagit.org/framasoft/mobilizon/-/blob/1.0.7/lib/mix/tasks/mobilizon/instance.ex#L132-133
+          genSecret =
+            "IO.puts(:crypto.strong_rand_bytes(64)" +
+            "|> Base.encode64()" +
+            "|> binary_part(0, 64))";
+
+          # Taken from here:
+          # https://github.com/elixir-lang/elixir/blob/v1.11.3/lib/mix/lib/mix/release.ex#L499
+          genCookie = "IO.puts(Base.encode32(:crypto.strong_rand_bytes(32)))";
+
+          evalElixir = "${launchers}/bin/mobilizon eval";
+        in
+        ''
+          set -euxo pipefail
+
+          if [ ! -f "${secretEnvFile}" ]; then
+            install -m 600 /dev/null "${secretEnvFile}"
+            cat > "${secretEnvFile}" <<EOF
+          # This file was automatically generated by mobilizon-setup-secrets.service
+          export MOBILIZON_AUTH_SECRET='$(${evalElixir} "${genSecret}")'
+          export MOBILIZON_INSTANCE_SECRET='$(${evalElixir} "${genSecret}")'
+          export RELEASE_COOKIE='$(${evalElixir} "${genCookie}")'
+          EOF
+          fi
+        '';
+
+      serviceConfig = {
+        Type = "oneshot";
+        User = user;
+        Group = group;
+        StateDirectory = "mobilizon";
+      };
+    };
+
+    # Add the required PostgreSQL extensions to the local PostgreSQL server,
+    # if local PostgreSQL is configured.
+    systemd.services.mobilizon-postgresql = mkIf isLocalPostgres {
+      description = "Mobilizon PostgreSQL setup";
+
+      after = [ "postgresql.service" ];
+      before = [ "mobilizon.service" "mobilizon-setup-secrets.service" ];
+      wantedBy = [ "mobilizon.service" ];
+
+      # Taken from here:
+      # https://framagit.org/framasoft/mobilizon/-/blob/1.1.0/priv/templates/setup_db.eex
+      script =
+        let
+          pgSu = "${pkgs.util-linux}/bin/runuser -u ${config.services.postgresql.superUser}";
+          psql = "${pgSu} -- ${postgresql}/bin/psql";
+        in
+        ''
+          ${psql} "${repoSettings.database}" -c "\
+            CREATE EXTENSION IF NOT EXISTS postgis; \
+            CREATE EXTENSION IF NOT EXISTS pg_trgm; \
+            CREATE EXTENSION IF NOT EXISTS unaccent;"
+        '';
+
+      serviceConfig.Type = "oneshot";
+    };
+
+    services.postgresql = mkIf isLocalPostgres {
+      enable = true;
+      ensureDatabases = [ repoSettings.database ];
+      ensureUsers = [
+        {
+          name = dbUser;
+          ensurePermissions = {
+            "DATABASE ${repoSettings.database}" = "ALL PRIVILEGES";
+          };
+        }
+      ];
+      extraPlugins = with postgresql.pkgs; [ postgis ];
+    };
+
+    users.users.${user} = {
+      description = "Mobilizon daemon user";
+      group = group;
+      isSystemUser = true;
```

Comment:
any particular reason you are using isSystemUser here?

Reply:
Yes, the mobilizon user is only useful to run the Mobilizon software. So by convention it should not have a UID >= 1000, which are reserved for "normal" users. Putting an explicit uid is a hassle, because you have to centralize uids to avoid collisions with other services, so to my understanding, `isSystemUser` is just a way to ask the script that generates user to generate a uid < 1000

Reply:
Just additional context around this.
For somebody managing an instance, there will be a problem around making backups.
the mobilizon user will be able to make backups, but it won't be able to upload them anywhere using ssh. In my experience for ssh to work properly, you have to have a `.ssh` directory containing the keys you want to use.
So somebody wanting to make backups, would either have to
- make root as a superuser make the dump and upload with ssh with the root user
- make the dump with the mobilizon user, then create another user that has a .ssh directory that will make the upload to an external service.
- make another user, make is as a superuser, make the dump and upload it with ssh to an external location.

I don't have a solution for this, i'm just bringing this up.

Reply:
If you are concerned about backups, you can also use the `services.postgresqlBackup` in Mobilizon's case. You could also use the `-i` option of `ssh` / `scp`, and every backup solution that can use SSH that I know of allows you to specify the path to a SSH key one way or another (for example, with rsync you can do `rsync -e "ssh -i /path/to/key" ...`). 

> My initial feeling is that, it might be better to make a user per service that you can login as.

I would strongly advise against that, I distinctly remember one of my teacher getting hacked precisely because someone could log as a user that should not have had a login shell / password.

> That way database permissions are neatly separated

I'm afraid I don't follow. Whether a user has a login shell and a password shouldn't change anything permission-wise with Postgresql, or am I missing something? I definitely agree that one generally should configure a user for each service, though.

> This has the assumption that uploading the backup to an external location will use ssh, maybe there is another way to do it and I'm just not seeing it.

I don't think SSH is the only way of uploading backups, I could see FTP being the standard way of doing backup in a company with a secure network environment and a somewhat old infrastructure. The way I see it though, if there is an "archiver" user with read access to every file to backup, it shouldn't have a login shell nor password, and if there is a need for a HOME, it could be `/var/lib/<my-archiver>` (I think this is quite common, even in some nixos services).

> One last consideration is that if they move in to a new instance, if they want to upload a dump, being able to login as that user would make it easier.

I've never used them myself, but I think Postgresql dumps take care of this. If you restore a database dumped with `pg_dump`, it should restore the proper permissions, so you can do it from the root / posgres user.

Reply:
Thanks a lot for the reply!

The problem I have with SSH, is that even though this allows a -i option to specify your private key, is that there is no way to specify the public key. It always assumes that your public key is living under `private_key.pub`. I haven't found a way to make ssh work for a user that doesn't have a `.ssh` directory under their home. Maybe that is just me though.

I didn't consider FTP, that might be the way.

| That way database permissions are neatly separated
Basically what I was saying is that if you don't do the backup with the service user that you have created, you have a need for a superuser (postgres or any other name you choose). It's the same thing with uploading a dump, you can do it from a superuser. It's not bad per se, but it forces you to manage a password on the superuser.

Just to try to summerize. The way you see it is
- Have a superuser (postgres or other) that will do the initial dump.
- For the backup script, this would be executed by a superuser that has a .ssh directory and that can upload the backup externally.

Let me check further on FTP!

Reply:
One last consideration regarding this
What about the remote iex shell on the node? I think this can be solved in a similar way by having a super user.
I wonder how to connect on the running service though.
Perhaps by providing the same environment variables as the service that is running and making a call to the package if it's in environment.systemPackages?

Reply:
I would discourage the use of FTP actually, I just used it as an example of another (maybe even older) way of doing backup ^^

The remote shell can be used without issue (I could test it) by any other user having access to the service, since it just uses inet sockets and the cookie for authentication

Reply:
Awesome. So as long as the COOKIE is the same, a user will be able to connect with a remote shell. great! Thanks for testing.